cmd/posix-list-dir_unix.go

```diff
@@ -110,13 +111,21 @@ var readDirBufPool = sync.Pool{
 }
 
 // Return all the entries at the directory dirPath.
-func readDir(dirPath string) (entries []string, err error) {
-	return readDirN(dirPath, -1)
+// leafFile is optional but if provided is used to
+// verify a given directory is actually a special
+// file. The suffix for such an entry is trimmed
+// and returned as if this is an actual file not
+// a directory.
+func readDir(dirPath string, leafFile string) (entries []string, err error) {
```

Comment:
Better not to bring leaf details inside readDir() as it affects other code like:

```
	fis, err := readDir(certsCAsDir, "")
```

leaf can should be done in `func (s *posix) ListDir()` instead.

Also stat on xl.json should be done once all the dir entries is read (to not cause randomio, disk entries will be in data blocks section and inodes will be in metadata blocks section on the disk)

Reply:
That would cause double entries maintained in memory @krishnasrinivas - the current implementation reduces the memory footprint as it filters at the source. I want to avoid the in-memory filtering as much as possible.

I don't know about random I/O - I think that is something I can't see conclusively in the tests. I will keep the code as is unless there is conclusive evidence.